src/pyspiffe/svid/x509_svid.py

```diff
@@ -1,34 +1,475 @@
 """
-This module manages X509Svid objects.
+This module manages X.509 SVID objects.
 """
 
-from typing import Mapping, Set
-from datetime import datetime
+import os
+
+import pem
+from typing import List, Union
+from typing.io import BinaryIO
+
+from cryptography import x509
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import ed25519, ed448, rsa, ec, dsa
+from cryptography.x509 import Certificate
+
 from pyspiffe.spiffe_id.spiffe_id import SpiffeId
 
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.serialization import load_der_private_key
+from cryptography.hazmat.primitives.serialization import load_pem_private_key
+
+from pyasn1.codec.der.decoder import decode
+
+from pyspiffe.svid.exceptions import (
+    InvalidLeafCertificateError,
+    InvalidIntermediateCertificateError,
+    ParseCertificateError,
+    ParsePrivateKeyError,
+    X509SvidError,
+    StoreCertificateError,
+    StorePrivateKeyError,
+    LoadCertificateError,
+    LoadPrivateKeyError,
+    normalize_exception_message,
+)
+
+_CERTS_FILE_MODE = 0o644
+_PRIVATE_KEY_FILE_MODE = 0o600
+
+_PRIVATE_KEY_TYPES = Union[
+    ed25519.Ed25519PrivateKey,
+    ed448.Ed448PrivateKey,
+    rsa.RSAPrivateKey,
+    dsa.DSAPrivateKey,
+    ec.EllipticCurvePrivateKey,
+]
+
 
 class X509Svid(object):
-    """Represents a SPIFFE X509 SVID."""
+    """
+    Represents a SPIFFE X.509-SVID.
+
+    Contains a SpiffeId, a private key and a chain of X.509 certificates.
+    """
 
     def __init__(
         self,
         spiffe_id: SpiffeId,
-        audience: Set[str],
-        expiry: datetime,
-        claims: Mapping[str, object],
-        token: str,
-    ):
-        """Creates a new X509Svid Object.
+        cert_chain: List[Certificate],
+        private_key: _PRIVATE_KEY_TYPES,
+    ) -> None:
+        """Creates a X509Svid instance.
+
+        Args:
+            spiffe_id: a SpiffeId instance.
+            cert_chain: a list representing a chain of X.509 Certificate.
+            private_key: a Private Key object.
+        """
+
+        self._spiffe_id = spiffe_id
+        self._cert_chain = cert_chain
+        self._private_key = private_key
+
+    def leaf(self) -> Certificate:
+        """Returns the Leaf X.509 certificate of the chain."""
+        return self._cert_chain[0]
```

Comment:
Is it safe to call `cert_chain[0]` without validating that there is at least on cert in the chain?

Reply:
It is safe because the methods that create an X509Svid enforce that the cert_chain have at least one Certificate, we can't have an X509Svid with an empty chain.

Reply:
The X509Svid constructor now enforce the invariants.